outlookmsgfile.py

```diff
@@ -240,50 +236,52 @@ def parse_properties(properties, is_top_level, container, doc):
         # Stream isn't present!
         logger.error("stream missing {}".format(streamname))
         continue
-      try:
-        value = tag_type.load(value, doc)
-      except KeyError as e:
-        logger.error("Error while reading stream: {} not found".format(str(e)))
-        continue
+
     else:
       # unrecognized type
       logger.error("unhandled property type {}".format(hex(property_type)))
       continue
 
-    ret[tag_name] = value
+    raw_properties[tag_name] = (tag_type, value)
 
-  # Post-processing: decode String8 strings using the code pages defined
-  # in properties.
-  #
-  # These properties are required to be present but in the real world they
-  # aren't always there, so there's a fallback to code page 1252 (Western
-  # Europe)
+  # Decode all FixedLengthValueLoader properties so we have codepage
+  # properties.
+  properties = { }
+  for tag_name, (tag_type, value) in raw_properties.items():
+    if not isinstance(tag_type, FixedLengthValueLoader): continue
+    try:
+      properties[tag_name] = tag_type.load(value)
+    except Exception as e:
+      logger.error("Error while reading stream: {}".format(str(e)))
 
-  # The encoding of the "BODY" (and HTML body) properties
-  if "PR_INTERNET_CPID" in ret:
-    body_codepage = code_pages.get(ret['PR_INTERNET_CPID'], 'cp1252')
-  else:
-    body_codepage = 'cp1252'
+  # String8 strings use code page information stored in other
+  # properties, which may not be present. Find the Python
+  # encoding to use.
 
-  # The encoding of "string properties of the message object"
-  if "PR_MESSAGE_CODEPAGE" in ret:
-    message_codepage = code_pages.get(ret['PR_MESSAGE_CODEPAGE'], 'cp1252')
-  else:
-    message_codepage = 'cp1252'
+  # The encoding of the "BODY" (and HTML body) properties.
+  body_encoding = None
+  if "PR_INTERNET_CPID" in properties and properties['PR_INTERNET_CPID'] in code_pages:
+    body_encoding = code_pages[properties['PR_INTERNET_CPID']]
 
-  for tag_name in tags_to_decode:
-    decoded = False
-    if tag_name == "BODY":
-      try:
-        ret[tag_name] = ret[tag_name].decode(body_codepage)
-        decoded = True
-      except UnicodeDecodeError:
-        pass
+  # The encoding of "string properties of the message object".
+  properties_encoding = None
+  if "PR_MESSAGE_CODEPAGE" in properties and properties['PR_MESSAGE_CODEPAGE'] in code_pages:
+    properties_encoding = code_pages[properties['PR_MESSAGE_CODEPAGE']]
+
+  # Decode all of the remaining properties.
+  for tag_name, (tag_type, value) in raw_properties.items():
+    if isinstance(tag_type, FixedLengthValueLoader): continue # already done, above
 
-    if not decoded:
-      ret[tag_name] = ret[tag_name].decode(message_codepage, errors='replace')
+    encoding = body_encoding if tag_name == "BODY" else properties_encoding
```

Comment:
This removes the "in-between" fallback from `body_encoding` to `properties_encoding` before the "don't know what else to try" fallback of `cp1252`. Is that intentional?

Reply:
Not intentional, thanks. I force-pushed a second attempt to get it right.